### PR TITLE
prov/gni: Fix NIC reference issue on MR

### LINKED ISSUE #1102: GNIX MR could grab references on two separate NICs
From #1086 

> >The NIC is only created if there are no NICs available. This would be the only NIC the registration could be tied to.

> If an EP was being created by another thread, you could have problems. There are bad assumptions being made here.

### PR DESCRIPTION
**Pull request by jswaro**
In a multi-threaded environment, it is possible that the NIC which
was allocated for the MR would not be the NIC picked by the memory
registration subsystem. To fix this, the NIC reference is taken later
during the registration step.

Signed-off-by: James Swaro <jswaro@cray.com>

@ztiffany 

closes #1102

### CODE REVIEW

**Comment by jswaro** [comment]
I am aware that there is still a threading issue with this PR. This PR is only addressing the reference count issue w.r.t. two separate NICs, not the usage of the ptag nic list lock. The fix for the threading will be a separate PR  as a resolution for #1098

**Comment by ztiffany** [comment]
Looks much cleaner! :+1:

**Comment by jswaro** [comment]
Merged content from #1099.

**Comment by jswaro** [comment]
closes #1086

File: prov/gni/src/gnix_mr.c [hunk 1]
```diff
@@ -311,31 +295,45 @@ static inline void *__gnix_generic_register(
...
+                        return NULL;
+                }
+        } else {
+                nic = dlist_first_entry(&gnix_nic_list_ptag[domain->ptag], 
+			struct gnix_nic, ptag_nic_list);
+                if (!nic)
+                        return NULL;
+		_gnix_ref_get(nic);	
+        }
+
```
Comment (chuckfossen) [inline prov/gni/src/gnix_mr.c:316]: This section has leading spaces instead of tabs.

Comment (chuckfossen) [reply prov/gni/src/gnix_mr.c:316]: hmm, I seem to be a minute late

**Pull request merged**